File: kotlin/app/src/main/java/com/example/cityexplorer/data/dtos/HexagonDtos.kt
package com.example.cityexplorer.data.dtos

import androidx.annotation.Keep
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class GetCountriesWithCitiesResponseDto(
    @SerialName("country")
    val country: String,

    @SerialName("cities")
    val cities: List<String>
)

@Serializable
data class GetHexagonsFromCityResponseDto(
    @SerialName("bbox")
    val bbox: List<Double>,

    @SerialName("hexagons")
    val hexagons: List<Hexagon> = emptyList()
)

@Serializable
data class GetPoisFromHexagonResponseDto(
    @SerialName("name")
    val name: String,

    @SerialName("type")
    val type: String,

    @SerialName("isPromoted")
    val isPromoted: Boolean
)

@Serializable
data class PostLocationBatchRequestDto(
    @SerialName("locations")
    val locations: List<CustomLocationStringTimestamp>
)

@Serializable
data class CustomLocationStringTimestamp(
    @SerialName("latitude")
    val latitude: Double,

    @SerialName("longitude")
    val longitude: Double,

    @SerialName("timestamp")
    val timestamp: String
)

@Serializable
data class CustomLocationDoubleTimestamp(
    @SerialName("latitude")
    val latitude: Double,

    @SerialName("longitude")
    val longitude: Double,

    @SerialName("timestamp")
    val timestamp: Long
)

@Serializable
data class PostLocationBatchResponseDto(
    @SerialName("updatedHexagons")
    val updatedHexagons: List<HexagonProgress>?,

    @SerialName("token")
    val token: String?
)

@Serializable
data class HexagonProgress(
    @SerialName("hexagonId")
    val hexagonId: String = "",

    @SerialName("progress")
    val progress: Double
)

@Serializable
data class Hexagon(
    @SerialName("id")
    val id: String,

    @SerialName("boundaries")
    val boundaries: List<List<Double>>,

    @SerialName("center")
    val center: List<Double>,

    @SerialName("weight")
    val weight: Double,

    @SerialName("progress")
    val progress: Double = 0.0
)

@Serializable
data class SelectedHexagon(
    @SerialName("weight")
    val weight: Double = 0.0,

    @SerialName("pois")
    val pois: List<GetPoisFromHexagonResponseDto>? = emptyList()
)


@Serializable
data class GenerateRouteRequestDto(
    @SerialName("userLatitude")
    val userLatitude: Double,

    @SerialName("userLongitude")
    val userLongitude: Double,

    @SerialName("duration")
    val duration: Int
)

@Serializable
data class GenerateRouteResponseDto(
    @SerialName("jobId")
    val jobId: String,

    @SerialName("token")
    val token: String?
)

@Keep
@Serializable
data class WorkerResult(
    @SerialName("route")
    val route: List<String> = emptyList()
)

File: csharp/api/Dtos/HexagonDtos.cs
using System.Text.Json.Serialization;

namespace csharp.Dtos;

public class GetCountriesWithCitiesResponseDto
{
    [JsonPropertyName("country")] public required string Country { get; init; }

    [JsonPropertyName("cities")] public required List<string> Cities { get; init; }
}

public class GetHexagonsFromCityRequestDto
{
    [JsonPropertyName("city")] public required string City { get; init; }
}

public class GetHexagonsFromCityResponseDto
{
    [JsonPropertyName("bbox")] public required List<double> Bbox { get; init; }

    [JsonPropertyName("hexagons")] public List<Hexagon> Hexagons { get; init; } = [];
}

public class GetPoisFromHexagonRequestDto
{
    [JsonPropertyName("hexagonId")] public required string HexagonId { get; init; }
}

public class GetPoisFromHexagonResponseDto
{
    [JsonPropertyName("name")] public required string Name { get; init; }

    [JsonPropertyName("type")] public required string Type { get; init; }

    [JsonPropertyName("isPromoted")] public required bool IsPromoted { get; init; }
}

public class PostLocationBatchRequestDto
{
    [JsonPropertyName("locations")] public required List<Location> Locations { get; init; }
}

public class Location
{
    [JsonPropertyName("latitude")] public double Latitude { get; init; }

    [JsonPropertyName("longitude")] public double Longitude { get; init; }

    [JsonPropertyName("timestamp")] public DateTime Timestamp { get; init; }
}

public class PostLocationBatchResponseDto
{
    [JsonPropertyName("updatedHexagons")] public List<HexagonProgress>? UpdatedHexagons { get; init; }

    [JsonPropertyName("token")] public string? Token { get; init; }
}

public class HexagonProgress
{
    [JsonPropertyName("hexagonId")] public string HexagonId { get; init; } = string.Empty;

    [JsonPropertyName("progress")] public double Progress { get; init; }
}

public class Hexagon
{
    [JsonPropertyName("id")] public required string Id { get; init; }

    [JsonPropertyName("boundaries")] public required List<List<double>> Boundaries { get; init; }

    [JsonPropertyName("center")] public required List<double> Center { get; init; }

    [JsonPropertyName("weight")] public double Weight { get; init; }
}

public class GenerateRouteRequestDto
{
    [JsonPropertyName("userLatitude")] public double UserLatitude { get; init; }

    [JsonPropertyName("userLongitude")] public double UserLongitude { get; init; }

    [JsonPropertyName("duration")] public int Duration { get; init; }
}

public class GenerateRouteResponseDto
{
    [JsonPropertyName("jobId")] public required string JobId { get; init; }

    [JsonPropertyName("token")] public string? Token { get; init; }
}

public class WorkerResult
{
    [JsonPropertyName("route")] public List<string> Route { get; init; } = [];
}